### ISSUE TITLE
Should conditional components be allowed in functions?

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#744]
**Reported by per&#x2e;ostlund on 23 May 2012 07:43 UTC**
Conditional components may only be used in connections, and connections are not allowed in functions. Is there any scenario where conditional components may be useful in functions, or should they be disallowed?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/744

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 23 May 2012 08:00 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 23 May 2012 08:27 UTC**
Components in a function are already free to be optionally included/removed by the compiler if they are not required to set the desired outputs, or if their value is never set, so making them conditional wouldn't add additional functionality.

That doesn't mean that the specification needs to state that they are disallowed in functions (or other specialised classes that can't contain connections) however, because they won't actually do anything, since variable values can only be set via connections.

**Comment by modelica-trac-importer** [comment]
**Comment by per&#x2e;ostlund on 23 May 2012 08:48 UTC**
Replying to [comment:2 pharman]:

> Components in a function are already free to be optionally included/removed by the compiler if they are not required to set the desired outputs, or if their value is never set, so making them conditional wouldn't add additional functionality.
> 
> That doesn't mean that the specification needs to state that they are disallowed in functions (or other specialised classes that can't contain connections) however, because they won't actually do anything, since variable values can only be set via connections.

I should perhaps have said that I'm mostly concerned about conditional formal parameters in functions, which changes a functions signature. E.g.:
```
function f
  input Real x if some condition;
end f;
```
I agree with what you say if one considers local variables in functions, but conditional formal parameters means that you have to evaluate the conditions to get the type of a function. It also makes it harder to use the function since a user need to know the conditions to be able to use the function. So if there is no reason to allow conditional formal parameters I think it would be good for both developers and users to disallow them.

I would argue that it's also good to disallow constructs that doesn't to anything, to help users write good code. But that's a different discussion.

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 23 May 2012 09:31 UTC**
Replying to [comment:3 per.ostlund@…]:
> I should perhaps have said that I'm mostly concerned about conditional formal parameters in functions, which changes a functions signature. E.g.:
> ```
> function f
>   input Real x if some condition;
> end f;
> ```
> I agree with what you say if one considers local variables in functions, but conditional formal parameters means that you have to evaluate the conditions to get the type of a function. It also makes it harder to use the function since a user need to know the conditions to be able to use the function.

I see, and agree that allowing conditional inputs in this way would only add complication from both the tool and user perspective, though my point on optimisation does also include inputs because they can have default values.

> I would argue that it's also good to disallow constructs that doesn't to anything, to help users write good code. But that's a different discussion.

...and if this was that discussion I would argue back that the language specification isn't for users, the tool giving the user sufficient information while they build the model is what helps them write good code.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 23 May 2012 09:33 UTC**
Although syntactically equivalent, I think your example would be semantically completely different from components. Components with a conditional attribute can be given a "not present" value (e.g. null). A formal input parameter that may or may not be present, is ... different. I can hardly see a useful application, but in essence you would give up the distinction between partially evaluated functions and other values. 

```mo
function f
  input Real x;
  input Real d if x < 10;
  output Real y = if x < 10 then d else x*2;
end f;
```

In such an example, what would f(time) be? a real value? a partially evaluated function?

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 23 May 2012 09:36 UTC**
Replying to [comment:4 pharman]:
> Replying to [comment:3 per.ostlund@…]:
> > I would argue that it's also good to disallow constructs that doesn't to anything, to help users write good code. But that's a different discussion.
> 
> ...and if this was that discussion I would argue back that the language specification isn't for users, the tool giving the user sufficient information while they build the model is what helps them write good code.

And I would argue, a language specification should ideally be **complete**. That is, every syntactically valid statement should be defined in some way (illegal or have a certain meaning). Leaving gaps is for sure not a good idea for a standardized language.

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 24 May 2012 09:24 UTC**
Replying to [comment:6 choeger]:
> And I would argue, a language specification should ideally be **complete**. That is, every syntactically valid statement should be defined in some way (illegal or have a certain meaning). Leaving gaps is for sure not a good idea for a standardized language.

I'm not advocating leaving gaps in the specification, I'm simply saying that the specification is there for the tools community not the user community who are engineers and scientists not computer-scientists.

It was my impression that it was already implied that conditional components were not allowed in functions due to the requirement that variables are set by connections, but Per pointed out a case in which they might. The 3.2 specification has in 4.4.5 the rather broad statement of _Other use of the component is illegal_, which is clearly open to interpretation.

The same example would be potentially an issue if the variable was declared in a model or a block as:
```
block b
  input Real x if some condition;
end b;
```
Setting this by an equation would be an error, however there is non-normative text stating _its modifier is ignored_, suggesting that such a variable could be set by a modification (and that modification can too include conditional variables that might have been removed). This doesn't fit with saying _Other use of the component is illegal_ when referring to the removal of connections. So I would suggest it is input-causality of a non-connector-variable that should be disallowed for conditional components, which would rule out both cases.

**Comment by modelica-trac-importer** [comment]
**Comment by per&#x2e;ostlund on 24 May 2012 09:35 UTC**
Replying to [comment:7 pharman]:
> The 3.2 specification has in 4.4.5 the rather broad statement of _Other use of the component is illegal_, which is clearly open to interpretation.

This has been fixed in the 3.2 rev1 specification.

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 24 May 2012 09:53 UTC**
I see a potential issue brewing here that I haven't seen discussed.  We are working toward (I don't know the current state of this, perhaps a solution was found) allowing blocks to be anonymously instantiated through function application syntax.  If we allow conditional components for blocks, but not for functions...then won't we have a problem (or at least a lot of confusion) when we allow function application syntax to implicitly instantiate things?  In other words, if blocks and functions are so different that one can have conditional components and the other cannot...then any proposal that attempts to treat them as conceptually similar could be dangerous.

I could be wrong since I haven't looked at this in a rigorous way, but it seems like these two proposals are on a collision course.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 24 May 2012 10:01 UTC**
Good point. We would need to add a special case disallowing this for blocks called as functions.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 24 May 2012 11:24 UTC**
That depends on if (and how) anonymous blocks would be specified. In my original draft, I proposed to instantiate them syntactically from the call-site. In that proposal, every formal parameter becomes a concrete component, and it would at least be possible to define the result. But this would of course open up some typing issues (and probably get us into trouble with unnamed arguments).

And btw: The idea was to treat them as conceptually different, but give a comfortable syntax to users.

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 24 May 2012 11:36 UTC**
But does a "comfortable syntax" necessarily mean "identical syntax".  For example, how about "invoke(SomeBlock,in1=...,in2=...)".  Keep in mind, I'm not proposing specific syntax, just that perhaps it might be better if the syntax was made explicitly distinct.  But I'm afraid I'm getting off topic for this ticket.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 24 May 2012 11:55 UTC**
No, not necessarily. That was just a pragmatic idea (we already have the grammar with a "semantic hole" at that point).

**Comment by modelica-trac-importer** [comment]
**Modified by sjoelund&#x2e;se on 18 Jun 2012 08:46 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 17 Oct 2012 16:23 UTC**
Since not handled at the last meeting rescheduling to the next one.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 20 Dec 2012 08:08 UTC**
Since not handled at the last meeting rescheduling to the next one.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Jun 2013 16:27 UTC**
Not urgent to fix - but seems no reason to have them.

**Comment by modelica-trac-importer** [comment]
**Changelog removed by hansolsson on 4 Jun 2013 16:27 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 11 Jun 2015 14:50 UTC**
Reconsidered and opened new ticket: #2057
It would still be possible to have an unused conditional variable in a function, but anything meaningful would be prohibited by that change.

Was going to open a new ticket, but since it is related and the solution works for both I add it here.

The specification in 4.4.5 states that conditional component with the condition being false is removed from the flattened DAE.

The issue is how valid the component must be when it is removed. 

The term 'removed' to me indicates that the component is first added (which implies that it would be valid to add it), and then 'removed'.

In particular this means:
* The type exists
* The modifier is legal (the example in the spec confirms that)
* A component with that type would not break restrictions on restricted classes (1)
* If the component is declared using 'outer' the corresponding 'inner' component should exist

(1): E.g. a block may not contain e.g. an Electrical pin with condition false.
\--
The proposal would be to clarify 4.4.5 to:

The component is first added to the flattened DAE regardless of the condition.
If the Boolean expression is false the component (including its modifier) is removed from the flattened DAE , and connections to/from the component are removed. 
[Adding the component and then removing it ensures that the component is valid.]
\--
Replaced that modifier "is ignored" by "is removed".

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 7 Feb 2017 14:32 UTC**
Replying to [comment:18 hansolsson]:
> Reconsidered and opened new ticket: #2057
> It would still be possible to have an unused conditional variable in a function, but anything meaningful would be prohibited by that change.
> 
> Was going to open a new ticket, but since it is related and the solution works for both I add it here.

As indicated a new ticket was opened, and closed with a correction.

Technically it is still possible to have a conditional component in a function - but I don't see how you can use it. Thus I view this ticket as low priority.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 1 Mar 2017 09:16 UTC**
Conclusion that no need to change specification (after #2057).
Agreement by acclamation.
